cpp/ql/lib/semmle/code/cpp/commons/Printf.qll

```diff
@@ -1046,39 +1060,64 @@ class FormatLiteral extends Literal {
-        )
+        len =
+          min(float cand |
+            // The first case handles length sub-specifiers
+            // Subtract one in the exponent because one bit is for the sign.
+            // Add 1 to account for the possible sign in the output.
+            cand = 1 + lengthInBase10(2.pow(this.getIntegralDisplayType(n).getSize() * 8 - 1))
+            or
+            // The second case uses range analysis to deduce a length that's shorter than the length
+            // of the number -2^31.
+            exists(Expr arg, float lower |
+              arg = this.getUse().getConversionArgument(n) and
+              lower = lowerBound(arg.getFullyConverted())
+            |
+              cand =
+                max(int cand0 |
+                  (
+                    // Include the sign bit in the length of `lower` if it can be negative
+                    if lower < 0
+                    then cand0 = 1 + lengthInBase10(lower.abs())
+                    else cand0 = lengthInBase10(lower)
+                  )
+                  or
+                  // We don't care about the sign of `upper`: if `upper` is negative, then we know
+                  // `lower` is also (possibly more) negative, and thus its length will be greater.
+                  cand0 = lengthInBase10(upperBound(arg.getFullyConverted()))
+                )
+            )
+          )
         or
         this.getConversionChar(n).toLowerCase() = "u" and
         // e.g. 2^32 - 1 = "4294967295"
-        exists(int sizeBits |
-          sizeBits =
-            min(int bits |
-              bits = this.getIntegralDisplayType(n).getSize() * 8
-              or
-              exists(IntegralType t |
-                t = this.getUse().getConversionArgument(n).getType().getUnderlyingType()
-              |
-                t.isUnsigned() and bits = t.getSize() * 8
-              )
-            ) and
-          len = (sizeBits / 10.0.log2()).ceil()
-          // convert the size from bits to decimal characters, and round up as you can't have
-          // fractional characters (10.0.log2() is the number of bits expressed per decimal character)
-        )
+        len =
+          min(float cand |
+            // The first case handles length sub-specifiers
+            cand = 2.pow(this.getIntegralDisplayType(n).getSize() * 8)
+            or
+            // The second case uses range analysis to deduce a length that's shorter than
+            // the length of the number 2^31 - 1.
+            exists(Expr arg, float lower |
+              arg = this.getUse().getConversionArgument(n) and
+              lower = lowerBound(arg.getFullyConverted())
+            |
+              cand =
+                max(float cand0 |
+                  (
+                    if lower < 0
+                    then
+                      // Calculate the value of `(unsigned)lower`.
+                      cand0 = 2.pow(any(IntType t | t.isUnsigned()).getSize() * 8) + lower
```

Comment:
I'm not convinced adding `+ lower` here is correct, as it's only a lower bound not the only achievable negative value.  I suspect the best thing to do here is `cand0 = 2.pow(any(IntType t | t.isUnsigned()).getSize() * 8)  - 1`

Reply:
Yeah, you're right. I had the logic confused.

Reply:
Fixed in 63f50a9eb7c1a9fe81d03931283c5e3702b1668f.